pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13463016?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13463016?buildTab=artifacts#/) on release-23.2.0-rc @ [a111ad6b18fb2d5c94af31cbf6018f61101fbc4c](https://github.com/cockroachdb/cockroach/commits/a111ad6b18fb2d5c94af31cbf6018f61101fbc4c):

Fatal error:

```
panic: connection has timed out &{0xc007eedf40 0xc00bc8d0a0 17a8a922889128240000000000000001 0xc00e47a5a0 { true} { false} 0}

```
Stack: 

```
goroutine 9483419 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0xc018399bc0, {0x786eba8, 0xc015535450})
	github.com/cockroachdb/cockroach/pkg/workload/schemachange/watch_dog.go:115 +0x25c
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 2133434
	github.com/cockroachdb/cockroach/pkg/workload/schemachange/watch_dog.go:133 +0x12b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
   "potentialExecErr": "0A000,42601,42704"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 0,
 "clientTimestamp": "11:08:58.88597",
 "ops": [
  "BEGIN",
  {
   "sql": "DROP SCHEMA \"public\" CASCADE",
   "expectedExecErr": "3F000"
  }
 ],
 "expectedExecErrors": "3F000",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: cannot drop schema \"public\" (SQLSTATE 3F000)"
}
{
 "workerId": 0,
 "clientTimestamp": "11:08:59.061757",
 "ops": [
  "BEGIN",
  {
   "sql": "CREATE TABLE public.table243 AS SELECT public.table226.\"col22 6_242\", public.table226.col226_237, public.table226.\"col226/_238\", public.table226.\"col22\\f6_228\", public.table226.col226😴_227, public.table226.col226_229, public.table226.col226_241, public.table226.col226_236, public.table226.col226̇_231, public.table226.\"col2%q26_232\", public.table226.\"col226/_238\", public.table226.\"col22 6_242\", public.table226.col226_229, public.table226.col226_240, public.table226.col226_233, public.table226.\"col22\\f6_228\", public.table226.col226_239, public.table226.col226_230, public.table226.\"co%ql22'6_234\" FROM public.table226, public.table226 FETCH FIRST 300000 ROWS ONLY",
   "expectedExecErr": "42701,42712"
  }
 ],
 "expectedExecErrors": "42701,42712",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: source name \"table226\" specified more than once (missing AS clause) (SQLSTATE 42712)"
}
{
 "workerId": 0,
 "clientTimestamp": "11:09:00.129431",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER DATABASE schemachange ADD REGION \"us-east3\"",
   "expectedExecErr": "42P12"
  }
 ],
 "expectedExecErrors": "42P12",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: cannot add region \"us-east3\" to database schemachange (SQLSTATE 42P12)"
}
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #117473 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-1 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35238

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13497496?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13497496?buildTab=artifacts#/) on release-23.2.0-rc @ [c8ffbdc4eeb3f656085edc33d6965b8f30cd7514](https://github.com/cockroachdb/cockroach/commits/c8ffbdc4eeb3f656085edc33d6965b8f30cd7514):

Fatal error:

```
panic: connection has timed out &{0xc008b0e140 0xc004a79060 17a9449b71ecab590000000000000001 0xc00a603b60 { true} { false} 0}

```
Stack: 

```
goroutine 17744 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0xc006d3bce0, {0x7867e08, 0xc005514640})
	github.com/cockroachdb/cockroach/pkg/workload/schemachange/watch_dog.go:115 +0x25c
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 9949
	github.com/cockroachdb/cockroach/pkg/workload/schemachange/watch_dog.go:133 +0x12b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: relation \"public.table141\" does not exist (SQLSTATE 42P01)"
}
{
 "workerId": 1,
 "clientTimestamp": "10:36:18.837062",
 "ops": [
  "BEGIN",
  {
   "sql": "DROP SCHEMA \"public\" CASCADE",
   "expectedExecErr": "3F000"
  }
 ],
 "expectedExecErrors": "3F000",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: cannot drop schema \"public\" (SQLSTATE 3F000)"
}
{
 "workerId": 1,
 "clientTimestamp": "10:36:18.893544",
 "ops": [
  "BEGIN",
  {
   "sql": "CREATE SEQUENCE IF NOT EXISTS public.seq142"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 1,
 "clientTimestamp": "10:36:18.966896",
 "ops": [
  "BEGIN",
  {
   "sql": "DROP SEQUENCE IF EXISTS public.seq140"
  },
  {
   "sql": "CREATE TYPE public.enum143 AS ENUM ('ddsaa', 'aa', 'sfda', 'sadf', 'ffsfds', 'as')"
  },
  {
   "sql": "ALTER DATABASE schemachange ADD REGION \"us-east3\"",
   "expectedExecErr": "42P12"
  }
 ],
 "expectedExecErrors": "42P12",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: cannot add region \"us-east3\" to database schemachange (SQLSTATE 42P12)"
}
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #117473 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-1 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/workload/schemachange/watch_dog.go
+++ b/pkg/workload/schemachange/watch_dog.go
@@ -55,16 +55,19 @@ func newSchemaChangeWatchDog(conn *pgxpool.Pool, logger *logger) *schemaChangeWa
 // when progress is detected.
 func (w *schemaChangeWatchDog) isConnectionActive(ctx context.Context) bool {
 	// Scan the session to make sure progress is being made first.
+	var status string
 	sessionInfo := w.conn.QueryRow(ctx,
-		"SELECT active_queries, kv_txn FROM crdb_internal.cluster_sessions WHERE session_id = $1",
+		"SELECT active_queries, kv_txn, status FROM crdb_internal.cluster_sessions WHERE session_id = $1",
 		w.sessionID)
 	lastTxnID := w.txnID
 	lastActiveQuery := w.activeQuery
-	if err := sessionInfo.Scan(&w.activeQuery, &w.txnID); err != nil {
+	if err := sessionInfo.Scan(&w.activeQuery, &w.txnID, &status); err != nil {
 		w.logger.logWatchDog(fmt.Sprintf("failed to get session information: %v\n", err))
 		return false
 	}
-	if w.activeQuery != lastActiveQuery {
+	if w.activeQuery != lastActiveQuery || status == "IDLE" {
+		// If the query has changed or the session is idle, it means the initial
+		// query has made progress.
 		return true
 	}
 	lastNumRetries := w.numRetries
@@ -112,7 +115,7 @@ func (w *schemaChangeWatchDog) watchLoop(ctx context.Context) {
 			}
 			totalTimeWaited += time.Second
 			if totalTimeWaited > maxTimeOutForDump {
-				panic(fmt.Sprintf("connection has timed out %v", w))
+				panic(fmt.Sprintf("connection has timed out; sessionID=%s activeQuery=%+v", w.sessionID, w.activeQuery))
 			}
 		}
 	}

